README.md
[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.8369746.svg)](https://doi.org/10.5281/zenodo.8369746)

This library provides functionalities to define and compose animated transitions on maps, which are then rendered in a web page.
The WebMercator map tiles required to show the transition are pre-loaded to ensure a smooth animation.
**npm** package link: [https://www.npmjs.com/package/map-transition-helper](https://www.npmjs.com/package/map-transition-helper).


## Table of Contents
- [Installation](#installation)
- [API Description](#api-description)
    - [Point Types](#point-types)
    - [Transition Functions](#transition-functions)
        - [Types](#transition-functions--types)
        - [Factory Functions](#transition-functions--factories)
        - [Composition Functions](#transition-functions--compositions)
        - [Predefined Transition Functions](#transition-functions--predefined)
    - [`preloadTransition` Function](#preloadtransition-function)
    - [Utility Functions](#utility-functions)
    - [Class `TileMap`](#class-tilemap)
        - [`constructor`](#class-tilemap--constructor)
        - [Properties](#class-tilemap--properties)
        - [Methods](#class-tilemap--methods)
        - [Events](#class-tilemap--events)
    - [Class `Tile`](#class-tile)
        - [`constructor`](#class-tile--constructor)
        - [Properties](#class-tile--properties)
        - [`clone(): Tile`](#class-tile--clone)
    - [Class `Frame`](#class-frame)
        - [`constructor`](#class-frame--constructor)
        - [Properties](#class-frame--properties)
        - [Methods](#class-frame--methods)
    - [Class `Transition`](#class-transition)
        - [`constructor`](#class-transition--constructor)
        - [Properties](#class-transition--properties)
        - [Methods](#class-transition--methods)
        - [Events](#class-transition--events)
    - [Enum `TransitionState`](#enum-transitionstate)
- [Usage Examples](#usage-examples)
- [Attribution](#attribution)


## Installation

The library is built using Rollup.
A demo with some event tests and transitions is available in [`demo/`](./demo).
The demo requires the library to be compiled.
To do so:

``` bash
$ npm install
$ npx rollup -c
```

To use the library in another project, either build it and use the generated ES module[^esm] (`lib/map-transition-helper.esm.js`) or the CommonJS build (`lib/map-transition-helper.umd.js`, global name `MapTransitionHelper`) directly.
Alternatively, it can be installed as an npm package, either from the command line:

``` bash
$ npm install map-transition-helper@0.4.4
```

or by adding the following to the `dependencies` property in the `package.json` of your project (with the appropriate version):

``` json
{
    ...
    "dependencies": {
        "map-transition-helper": "0.4.4",
        ...
    },
    ...
}
```

[^esm]: The ES module build expects some external dependencies to exist, namely the D3.js modules it uses internally.
Unfortunately, I have not yet gotten the ES build to load in the browser without it complaining about the external imports.
However, it can be used with a build tool like WebPack or Rollup.


## API Description

### Point Types

Different types of points exist throughout the API:
Two- and three-dimensional cartesian points, two-dimensional geographical coordinates, and three-dimensional geographical view points that include a zoom level.

``` typescript
interface Point2D {
  x: number;
  y: number;
}

interface Point3D {
  x: number;
  y: number;
  z: number;
}

interface Coordinate {
  lat: number;
  lng: number;
}

interface ViewPoint {
  lat: number;
  lng: number;
  zoom: number;
}
```


### <a id="transition-functions"></a>Transition Functions

Transitions can be defined by a function that, for each point in time, returns a [`ViewPoint`](#class-viewpoint); that is, a tuple of latitude, longitude, and zoom level.
Transitions can be chained, and the time (in [0,1]) they receive can be pre-affected by an easing function.

#### <a id="transition-functions--types"></a>Types

##### `EasingFunction`

``` typescript
type EasingFunction = (t: number) => number;
```

An easing function takes a value `t` between 0 and 1, and returns a value (ideally between) 0 and 1.
Easing functions are used to define how a transition should behave regarding speed.

##### `PanningFunction`

``` typescript
type PanningFunction = (t: number) => Coordinate;
```

A panning function is a function that returns a coordinate for each point in time between the start time `t`=0 and the end time `t`=1.

##### `ZoomFunction`

``` typescript
type ZoomFunction = (t: number) => number;
```

A zoom function is a function that returns a zoom level for each point in time between the start time `t`=0 and the end time `t`=1.

##### `TransitionFunction`

``` typescript
type TransitionFunction = (t: number) => ViewPoint;
```

A transition function returns an entire `ViewPoint` (`Coordinate` plus zoom level) for each point in time between the start time `t`=0 and the end time `t`=1.


#### <a id="transition-functions--factories"></a>Factory Functions

The factory functions can be used to create basic building blocks for more complex transition functions.
Very custom functions can be written instead, of course.


##### `createPanning(p0: Coordinate, p1: Coordinate): PanningFunction`

Create a linear panning function between two coordinates.
The transition will be linear in the (Mercator) image space.

| Parameter | Type | Description |
| --- | --- | --- |
| `p0` | `Coordinate` | Start coordinate |
| `p1` | `Coordinate` | End coordinate |


##### `createPanningGreatCircle(p0: Coordinate, p1: Coordinate): PanningFunction`

Create a great circle panning function between two points.
The resulting panning will happen along the geodesic line between `p0` and `p1`.

| Parameter | Type | Description |
| --- | --- | --- |
| `p0` | `Coordinate` | Start coordinate |
| `p1` | `Coordinate` | End coordinate |


##### `createZoom(z0: number, z1: number): ZoomFunction`

Create a linear zoom function between two zoom levels.
This is just a thin wrapper around [`d3.interpolateNumber`](https://github.com/d3/d3-interpolate#interpolateNumber).

| Parameter | Type | Description |
| --- | --- | --- |
| `z0` | `number` | Start zoom |
| `z1` | `number` | End zoom |


#### <a id="transition-functions--compositions"></a>Composition Functions

Basic building blocks, but also whole transition functions, can be composed for more complex behavior.
This can happen multiple times.


##### `addPanningEase(pan: PanningFunction, ease: EasingFunction): PanningFunction`

Add an easing function to a panning function.
The result will be `t -> pan(ease(t))`.

| Parameter | Type | Description |
| --- | --- | --- |
| `pan` | `PanningFunction` | Initial panning function |
| `ease` | `EasingFunction` | Easing function to apply |


##### `addZoomingEase(zoom: ZoomFunction, ease: EasingFunction): ZoomFunction`

Add an easing function to a zoom function.
The result will be `t -> zoom(ease(t))`.

| Parameter | Type | Description |
| --- | --- | --- |
| `zoom` | `ZoomFunction` | Initial zoom function |
| `ease` | `EasingFunction` | Easing function to apply |


##### `addTransitionEase(trans: TransitionFunction, ease: EasingFunction): TransitionFunction`

Add an easing function to a transition function.
The result will be `t -> trans(ease(t))`.

| Parameter | Type | Description |
| --- | --- | --- |
| `trans` | `TransitionFunction` | Initial transition function |
| `ease` | `EasingFunction` | Easing function to apply |


##### `createTransitionFunction(pan: PanningFunction, zoom: ZoomFunction): TransitionFunction`

Join a panning and a zoom function to create a transition function.

| Parameter | Type | Description |
| --- | --- | --- |
| `pan` | `PanningFunction` | Panning function |
| `zoom` | `ZoomFunction` | Zoom function |


##### <a id="function-jointransitions"></a>`joinTransitions(transs: Array<TransitionFunction>, weights?: Array<number>): TransitionFunction`

Join multiple transition functions together.
Returns a single transition function from `t`=0 to `t`=1.
The portions of the new time domain the passed transition functions occupy is given by the `weights` array.
If that is not passed, all functions will have equal weight.

| Parameter | Type | Description |
| --- | --- | --- |
| `transs` | `Array<TransitionFunction>` | Array of transition functions |
| `weights` | `Array<number>` (optional) | Array of time portion weights |


#### <a id="transition-functions--predefined"></a>Predefined Transition Functions

Some basic transition functions come predefined.

##### `createLinearPanTransition(p0: ViewPoint, p1: ViewPoint): TransitionFunction`

Create a linear panning transition without zooming.
The zoom level of the first view point will be used throughout.

| Parameter | Type | Description |
| --- | --- | --- |
| `p0` | `ViewPoint` | Start point |
| `p1` | `ViewPoint` | End point |


##### `createLinearZoomAndPanTransition(p0: ViewPoint, p1: ViewPoint): TransitionFunction`

Create a linear pan and zoom.
Can also be used to just zoom.

**NOTE:** In some cases, this is probably not what you want:
Zooming is linear in the zoom level, but panning speed then needs to scale with the current
zoom level so that it is perceived as linear.
For zoom+pan transitions, use [`createPerceivedLinearZoomAndPanTransition`](#createperceivedlinearzptransition) instead.

| Parameter | Type | Description |
| --- | --- | --- |
| `p0` | `ViewPoint` | Start point |
| `p1` | `ViewPoint` | End point |


##### <a id="createperceivedlinearzptransition"></a>`createPerceivedLinearZoomAndPanTransition(p0: ViewPoint, p1: ViewPoint): TransitionFunction`

Create a pan and zoom transition that is perceived as being linear.
The panning will depend on the zoom level.

| Parameter | Type | Description |
| --- | --- | --- |
| `p0` | `ViewPoint` | Start point |
| `p1` | `ViewPoint` | End point |


##### `createBoxTransition(p0: ViewPoint, p1: ViewPoint, canvasSize: Point2D, minZoom?: number): TransitionFunction`

Create a box transition: zoom out from `p0` until `p1` is visible.
Then pan to `p1`.
Then zoom into `p1`.

| Parameter | Type | Description |
| --- | --- | --- |
| `p0` | `ViewPoint` | Start point |
| `p1` | `ViewPoint` | End point |
| `canvasSize` | `Point2D` | Size of the canvas the transition will run on |
| `minZoom` | `number` (optional) | Zoom level for panning phase. If not given, will be calculated to fit. |


##### `createTriangularTransition(p0: ViewPoint, p1: ViewPoint, canvasSize: Point2D, minZoom?: number): TransitionFunction`

Create a (perceived) triangular transition:
Zoom out and pan from `p0` to the midpoint between `p0` and `p1` in the image space, at which time both `p0` and `p1` will be visible.
Then, continue panning and zooming in again until centered on `p1`.

| Parameter | Type | Description |
| --- | --- | --- |
| `p0` | `ViewPoint` | Start point |
| `p1` | `ViewPoint` | End point |
| `canvasSize` | `Point2D` | Size of the canvas the transition will run on |
| `minZoom` | `number` (optional) | Zoom level for panning phase. If not given, will be calculated to fit. |


##### `createVanWijkAndNuijTransition(p0: ViewPoint, p1: ViewPoint, canvasSize: Point2D, rho: number = 1.4): TransitionFunction`

Create a transition following the hyperbolic path recommended by van Wijk and Nuij: *"Smooth and efficient zooming and panning"* (Proc. InfoVis, 2003).

**Note:** As of now, this library only supports Mercator projection.
This transition should be done in image space, even if the `TransitionFunction` outputs geographical space ViewPoints.
For now, this function will assume Mercator projection is used for its internal calculations.
This should be amended once different projections are supported.

Create a (perceived) triangular transition:
Zoom out and pan from `p0` to the midpoint between `p0` and `p1` in the image space, at which time both `p0` and `p1` will be visible.
Then, continue panning and zooming in again until centered on `p1`.

| Parameter | Type | Description |
| --- | --- | --- |
| `p0` | `ViewPoint` | Start point |
| `p1` | `ViewPoint` | End point |
| `canvasSize` | `Point2D` | Size of the canvas the transition will run on |
| `rho` | `number` (optional) | Parameter ρ, which determines the shape of the hyperbolic path in (u, w) space (see the paper). If not passed, the recommended value of 1.4 is used. |


### `preloadTransition` Function

``` typescript
async function preloadTransition(
  transitionFunction: TransitionFunction,
  canvasSize: Point2D,
  numberOfFrames: number,
  tileMap?: TileMap
): Promise<Transition>
```

This is the central functionality of the library.
It preloads all map tiles and pre-calculates all [`Frame`s](#class-frame) needed for a transition, and returns a [`Transition`](#class-transition) object asynchronously.

| Parameter | Type | Description |
| --- | --- | --- |
| `transitionFunction` | `TransitionFunction` | Transition function which describes the position of the viewing window and its zoom for every point during the transition. |
| `canvasSize` | `Point2D` | size of the canvas. |
| `numberOfFrames` | `number` | number of frames of the transition (directly determines the length of the animation). On most devices, 60 frames are displayed per second. |
| `tileMap` | `TileMap`  (optional) | If set by the user, this tilemap object is used to store the bitmaps of the needed Mercator tiles. Can be useful if more than one transition should be stored in the tilemap. This can lead to less memory consumption since a tile may participate in more than one transition. |


### Utility Functions

#### `calcMinZoomCenteredOn(center: Coordinate, outer: Coordinate, canvasSize: Point2D, margin?: number): number`

Calculate the minimal zoom level, centered on `center` on a canvas of size `canvasSize`, such that `outer` is visible `margin` pixels from the canvas edge.
Returns the minimal (strictly: maximal) zoom level that satisfies the condition.

| Parameter | Type | Description |
| --- | --- | --- |
| `center` | `Coordinate` | Center of the map |
| `outer` | `Coordinate` | Other point that should be visible |
| `canvasSize` | `Point2D` | Size of the canvas |
| `margin` | `number` (optional) | Distance of `outer` from the edge in pixels, default 20 |


#### `calcZoomFitPoints(coordinates: Array<Coordinate>, canvasSize: Point2D, margin: number = fitMargin,): number`

Calculate the minimal (maximal) zoom level such that all `coordinates` are still visible in a map with size `canvasSize`, with all points at least `margin` pixels from the edge of the canvas.

| Parameter | Type | Description |
| --- | --- | --- |
| `coordinates` | `Array<Coordinate>` | Array of coordinates that should be visible |
| `canvasSize` | `Point2D` | Size of canvas |
| `margin` | `number` (optional) | Minimum distance of coordinates from the edge of the canvas in pixels, default 20 |


#### `calculateFrame(canvasSize: Point2D, currentPoint: ViewPoint): Frame`

Calculate the [`Frame`](#class-frame) object that describes a particular [`ViewPoint`](#class-viewpoint) for a given canvas size.


### Class `TileMap`

An object of this class preloads and stores the map tiles required for transitions.
`TileMap` objects can be reused for multiple transitions.
`TileMap` extends `EventTarget` and fires events when tiles are loaded, fail to load, and loading is complete.


#### <a id="class-tilemap--constructor"></a>`constructor`

``` typescript
constructor(
    readonly _tileServerUrl: string = 'https://services.arcgisonline.com/arcgis/rest/services/World_Imagery/MapServer/tile/{z}/{y}/{x}',
    readonly _maxZoom: number = 19,
)
```

#### <a id="class-tilemap--properties"></a>Properties

##### `numTiles: number`

The number of tiles stored in the tile map.


#### <a id="class-tilemap--methods"></a>Methods

##### `hasTile(tile: Tile): boolean`

Returns `true` if the tile described by `tile` is loaded.

##### `getProgress(): { loading: number, finished: number, failed: number, total: number }`

Get the current loading progress, for example for a progress bar.
`total` is the sum of `loading`, failed, and `finished`.

##### `async addToTileMap(tile: Tile): Promise<void>`

Preload `tile`, if it has not been loaded yet.

##### `getTileImage(tile: Tile): ImageBitmap`

Get the image data for the passed tile coordinates.
If the coordinates are outside the world boundaries, returns a white tile instead.

##### `async fetchFailedTiles(): Promise<void>`

Try again to load all tiles that failed loading.

#### <a id="class-tilemap--events"></a>Events

##### `tileadded`

Fired each time a new (unique) tile is added.
The `CustomEvent`'s `detail` property contains the `Tile` instance.

##### `tileloaded`

Fired each time a tile is successfully loaded.
The `CustomEvent`'s `detail` property contains the `Tile` instance.

##### `tilefailed`

Fired each time a tile load fails.
The `CustomEvent`'s `detail` property contains the `Tile` instance.

##### `loaded`

Fired each time all queued tiles are loaded.


### Class `Tile`

Class which represents a map tile with x, y an zoom value.

#### <a id="class-tile--constructor"></a>`constructor`

``` typescript
constructor(
  readonly x: number,
  readonly y: number,
  readonly zoom: number
)
```

Does not usually need to be called directly.
This is handled by `preloadTransition`.

#### <a id="class-tile--properties"></a>Properties

The properties are the same as the parameters of the constructor.

#### <a id="class-tile--clone"></a>`clone(): Tile`

Create a new `Tile` instance with the same `x`, `y` and `zoom` values.


### Class `Frame`

Frame class where all necessary information about a frame are stored.
Especially the required tiles for a frame are saved here.

#### <a id="class-frame--constructor"></a>`constructor`

``` typescript
constructor(
  tiles: Array<[number, number, number]>,
  readonly offset: Point2D,
  readonly translation: Point2D,
  readonly canvasSize: Point2D,
  readonly scale: number,
  readonly zoom: number
)
```

| Parameter | Type | Description |
| --- | --- | --- |
| `tiles` | `Array<[number, number, number]>` | tile coordinates that are required for a frame, as produced by d3-tile |
| `offset` | `Point2D` | offset of the canvas center from the WebMercator origin |
| `translation` | `Point2D` | d3-tile translation, used to align the tile quadtree to the projection |
| `canvasSize` | `Point2D` | the size of the corresponding canvas (x=width, y=height) |
| `scale` | `number` | scale value for the tiles in this frame. Only needed for drawing the tiles into the canvas |
| `zoom` | `number` | the current zoom level |

Does not usually need to be called directly.
This is handled by `preloadTransition`.

#### <a id="class-frame--properties"></a>Properties

The properties are the same as the parameters of the constructor.

#### <a id="class-frame--methods"></a>Methods

##### <a id="class-frame--project"></a>`project(coord: Coordinate | ViewPoint): Point2D`

Calculate the pixel coordinates of real-world coordinates.
This is the inverse of [`unproject`](#class-frame--unproject).

##### <a id="class-frame--unproject"></a>`unproject(coord: Point2D): ViewPoint`

Calculate the latitude, longitude and zoom level shown at the pixel coordinate `coord`.
This is the inverse of [`project`](#class-frame--project).

##### `isCoordinateVisible(c: Coordinate, offset: number = 0): boolean`

Checks if the coordinate is within the drawable area of the canvas in the frame.
Optionally, an offset can be passed, similar to the one specified in [`borderPosition`](#class-frame--borderposition).
The offset is perpendicular to the border and positive inwards of the frame border.


##### <a id="class-frame--borderposition"></a>`borderPosition(c: Coordinate | ViewPoint, offset: number = 0): { x, y, border, direction }`

Calculate the position `c` at the border of the frame that is in the direction of a coordinate.
This can be used for positions that are visible in the current frame as well, but should not be.
An optional parameter `offset` specifies the perpendicular distance of the point to the border it should lie on, with positive values going into the frame.
This can be used to place objects of known size at the frame border in different alignments.
Returns an object with the x and y coordinates relative to the frame, the border (top, left, bottom, or right) at which the shortest path to the coordinate leaves the frame, and the direction (in radians) towards the point from the center of the frame.



### Class `Transition`

Main class to interact with.
This extends `EventTarget`, so `addEventListener` is supported.


#### <a id="class-transition--constructor"></a>`constructor`

``` typescript
constructor(frames: Frame[], tileMap: TileMap)
```

Takes two parameters, an array of [`Frame`s](#class-frame) and a [`TileMap`](#class-tilemap).
The constructor need never be called directly.
Use the [`preloadTransition`](#preloadtransition-function) function instead.

#### <a id="class-transition--properties"></a>Properties

##### `frameArray: Frame[]`

Property (read-only) containing all [`Frame`s](#class-frame).

##### `length: number`

Property (read-only) containing the number of [`Frame`s](#class-frame) in the transition.

##### `tileMap: TileMap`

Read-write property containing the transition's [`TileMap`](#class-tilemap).

##### `currentFrame: Frame`

Read-only property containing the current [`Frame`](#class-frame).

##### `currentFrameIndex: number`


CHANGELOG
                             Release Notes
                             -------------

2023-09-26                                                        v0.5.0

Added the hyperbolic transition path suggested by van Wijk and Nuij
(2003) to the pre-defined transitions.


2023-03-17                                                        v0.4.4

Add an optional offset to the visibility calculation of coordinates in a
frame. Remove unneeded dependencies.


2023-03-13                                                        v0.4.3

Fix a bug with the border calculation for border positions (#13).


2023-03-10                                                        v0.4.2

Add an optional offset to the calculation of border positions towards
coordinates outside of the visible area (contributor: Markus Stengel).


2023-03-09                                                        v0.4.1

Fixed `Transition.cancel()` stepping one frame after cancellation. Added
a convenience method to calculate the border position of a frame towards
a coordinates outside of the visible area.


2023-01-13                                                        v0.4.0

Fixed visibility checks for coordinates in the `Frame` class. Switched
the names of the functions `project` and `unproject` in the `Frame`
class to better align with their respective behavior.


2022-12-29                                                        v0.3.2

First public release with stable API and documentation. Fixed some last
exceptions in edge cases, such as trying to start a newly-finished
transition. Fixed blurry tile rendering.


; vim: tw=72


demo/main.js
const {
  createPanning,
  createPanningGreatCircle,
  createZoom,
  preloadTransition,
  joinTransitions,
  TileMap,
  createTransitionFunction,
  createLinearPanTransition,
  createPerceivedLinearZoomAndPanTransition,
  createLinearZoomAndPanTransition,
  createBoxTransition,
  createTriangularTransition,
  createVanWijkAndNuijTransition,
} = MapTransitionHelper;

const canvas = document.querySelector('canvas');

const transitionEventTypes = ['frame', 'render', 'play', 'pause', 'cancel', 'finish'];
const tilemapEventTypes = ['tileadded', 'tileloaded', 'tilefailed', 'loaded'];
const eventTypes = [...transitionEventTypes, ...tilemapEventTypes];
const eventIndicators = eventTypes.map((evt) => document.querySelector(`#event-${evt}`));
const lastFrame = eventTypes.map((_) => -30);
let currentFrameIndex = 0;

const tileMap = new TileMap(`https://tile.openstreetmap.org/{z}/{x}/{y}.png`);

tilemapEventTypes.forEach((evt) => {
  const i = eventTypes.indexOf(evt);
  tileMap.addEventListener(evt, (_) => (lastFrame[i] = currentFrameIndex));
});

function updateTilemapStatus() {
  const { loading, finished, failed, total } = tileMap.getProgress();
  document.querySelector(
    '#tilemap-progress',
  ).innerText = `${loading}loa+${finished}fin+${failed}fail / ${total}`;
  requestAnimationFrame(updateTilemapStatus);
}
requestAnimationFrame(updateTilemapStatus);

// places
const stuttgart1 = { label: 'stuttgart', lat: 48.7449, lng: 9.1048, zoom: 17 };
const stuttgart2 = { ...stuttgart1, zoom: 12 }; // zoomed out
const pforzheim = { label: 'pforzheim', lat: 48.8905, lng: 8.7037, zoom: 12 };
const germersheim = { label: 'germersheim', lat: 49.222, lng: 8.3661, zoom: 12 };
const frankfurt1 = { label: 'frankfurt', lat: 50.1013, lng: 8.6648, zoom: 12 };
const frankfurt2 = { ...frankfurt1, zoom: 3 };
const newYork1 = { label: 'new york', lat: 40.7084, lng: -74.0176, zoom: 3 };
const newYork2 = { ...newYork1, zoom: 7 };
const philadelphia = { label: 'philadelphia', lat: 39.9417, lng: -75.1571, zoom: 13 };
const tokyo = { label: 'Tōkyō', lat: 35.689, lng: 139.692, zoom: 8 };
const berlin = { label: 'Berlin', lat: 52.52, lng: 13.405, zoom: 8 };

const places = [
  stuttgart1,
  pforzheim,
  germersheim,
  frankfurt1,
  newYork1,
  philadelphia,
  tokyo,
  berlin,
];

// intermediate transitions
const zoomOutStuttgart = createTransitionFunction(
  createPanning(stuttgart1, stuttgart2),
  createZoom(stuttgart1.zoom, stuttgart2.zoom),
);
const moveToFrankfurt = joinTransitions(
  [
    createLinearPanTransition(stuttgart2, pforzheim),
    createLinearPanTransition(pforzheim, germersheim),
    createLinearPanTransition(germersheim, frankfurt1),
  ],
  [1, 1.12, 0.8],
);
const zoomOutFrankfurt = createTransitionFunction(
  createPanning(frankfurt1, frankfurt2),
  createZoom(frankfurt1.zoom, frankfurt2.zoom),
);
const moveToNewYork = createLinearPanTransition(frankfurt2, newYork1);
const zoomInNewYork = createTransitionFunction(
  createPanning(newYork1, newYork2),
  createZoom(newYork1.zoom, newYork2.zoom),
);
const moveToPhiladelphia = createPerceivedLinearZoomAndPanTransition(newYork2, philadelphia);
const moveBackToNewYork = createPerceivedLinearZoomAndPanTransition(philadelphia, newYork2);

const concatenatedTransition = joinTransitions(
  [
    zoomOutStuttgart,
    moveToFrankfurt,
    zoomOutFrankfurt,
    moveToNewYork,
    zoomInNewYork,
    moveToPhiladelphia,
    moveBackToNewYork,
  ],
  [1, 3, 1, 2.5, 1.2, 2.3, 1.8],
);

const linearTransition = createLinearPanTransition(stuttgart2, frankfurt1);
const boxTransition = createBoxTransition(stuttgart2, frankfurt1, {
  x: canvas.width,
  y: canvas.height,
});
const triangularTransition = createTriangularTransition(stuttgart2, frankfurt1, {
  x: canvas.width,
  y: canvas.height,
});
const aroundTheWorld = joinTransitions([
  createLinearPanTransition({ lat: 0, lng: 0, zoom: 3 }, { lat: 0, lng: 120 }),
  createLinearPanTransition({ lat: 0, lng: 120, zoom: 3 }, { lat: 0, lng: -120 }),
  createLinearPanTransition({ lat: 0, lng: -120, zoom: 3 }, { lat: 0, lng: 0 }),
]);
const vanWijkAndNuij = createVanWijkAndNuijTransition(
  stuttgart1,
  newYork2,
  {
    x: canvas.width,
    y: canvas.height,
  },
  1.5,
);
let transition;
requestAnimationFrame(async (_) => {
  document.body.toggleAttribute('inert', true);
  transition = await preloadTransition(
    vanWijkAndNuij,
    { x: canvas.width, y: canvas.height },
    300,
    tileMap,
  );
  registerEventListeners(transition);
  requestAnimationFrame(() => {
    document.querySelector('#initialize')?.click();
  });
  document.body.toggleAttribute('inert', false);
});

const methodSelect = document.querySelector('#transition-method');
methodSelect.addEventListener('input', (evt) => {
  let promise;

  switch (evt.target.value) {
    case 'concatenated':
      promise = preloadTransition(
        concatenatedTransition,
        { x: canvas.width, y: canvas.height },
        1200,
        tileMap,
      );
      break;
    case 'linear':
      promise = preloadTransition(
        linearTransition,
        { x: canvas.width, y: canvas.height },
        180,
        tileMap,
      );
      break;
    case 'box':
      promise = preloadTransition(
        boxTransition,
        { x: canvas.width, y: canvas.height },
        180,
        tileMap,
      );
      break;
    case 'triangle':
      promise = preloadTransition(
        triangularTransition,
        { x: canvas.width, y: canvas.height },
        180,
        tileMap,
      );
      break;
    case 'around world':
      promise = preloadTransition(
        aroundTheWorld,
        { x: canvas.width, y: canvas.height },
        180,
        tileMap,
      );
      break;
    case 'van Wijk and Nuij':
      promise = preloadTransition(
        vanWijkAndNuij,
        { x: canvas.width, y: canvas.height },
        300,
        tileMap,
      );
      break;
    default:
      throw new Error(`unknown transition method: ${evt.target.value}`);
  }

  document.body.toggleAttribute('inert', true);
  promise.then((t) => {
    transition.cancel();
    requestAnimationFrame((_) => {
      canvas
        .getContext('2d')
        .clearRect(-canvas.width / 2, -canvas.height / 2, canvas.width, canvas.height);
      transition = t;
      registerEventListeners(transition);
      document.body.toggleAttribute('inert', false);
    });
  });
});

document.querySelector('#fetch-failed').addEventListener('click', async (_) => {
  document.body.toggleAttribute('inert', true);
  await tileMap.fetchFailedTiles();
  document.body.toggleAttribute('inert', false);
});

document
  .querySelector('button#initialize')
  .addEventListener('click', (_) => transition?.initialize(canvas.getContext('2d')));
document.querySelector('button#play').addEventListener('click', (_) => transition?.play());
document.querySelector('button#pause').addEventListener('click', (_) => transition?.pause());
document.querySelector('button#cancel').addEventListener('click', (_) => transition?.cancel());

function registerEventListeners(transition) {
  transitionEventTypes.forEach((evt, i) => {
    transition.addEventListener(evt, (_) => (lastFrame[i] = currentFrameIndex));
  });

  transition.addEventListener('render', (e) => checkVisibility(e, transition.currentFrame));
}

function updateEventIndicators() {
  for (let i = 0; i < eventTypes.length; ++i) {
    const distance = currentFrameIndex - lastFrame[i];
    const falloffPercent = (100 * Math.min(distance, 30)) / 30;
    eventIndicators[i].style.filter = `grayscale(${falloffPercent.toFixed(0)}%)`;
  }

  currentFrameIndex++;
  requestAnimationFrame(updateEventIndicators);
}
requestAnimationFrame(updateEventIndicators);

const currentFrame = document.querySelector('#current-frame-nr');
function updateCurrentFrameLabel() {
  if (transition) {
    const fnr = transition.currentFrameIndex;
    const s = `Current: ${fnr}`;
    if (s !== currentFrame.innerHTML) currentFrame.innerHTML = s;
  }

  requestAnimationFrame(updateCurrentFrameLabel);
}
requestAnimationFrame(updateCurrentFrameLabel);

const frameSetInput = document.querySelector('#frame-nr-input');
document.querySelector('#submit-frame-nr').addEventListener('click', (_) => {
  const fnr = frameSetInput.valueAsNumber;
  if (transition) transition.currentFrameIndex = fnr;
});

const cursorPosition = document.querySelector('#cursor-position');
canvas.addEventListener('mouseleave', (_) => {
  cursorPosition.innerHTML = `outside canvas`;
});
requestAnimationFrame((_) => {
  const { top: canvasTop, left: canvasLeft } = canvas.getBoundingClientRect();
  function updateCursorPosition(evt) {
    if (!transition) return;

    const { layerX, layerY } = evt;
    const x = Math.round(layerX - canvasLeft);
    const y = Math.round(layerY - canvasTop);

    const frame = transition.getFrame(transition.currentFrameIndex);
    if (frame === undefined) return;

    const viewPoint = frame.unproject({ x, y });

    cursorPosition.innerHTML = `
      x: ${x}px,
      y: ${y}px,
      lat: ${viewPoint.lat.toFixed(5)},
      lng: ${viewPoint.lng.toFixed(5)},
      zoom: ${viewPoint.zoom.toFixed(2)}
    `;
  }
  canvas.addEventListener('mouseenter', updateCursorPosition);
  canvas.addEventListener('mousemove', updateCursorPosition);
});

// test visibility of some positions in the frame
function checkVisibility(evt, frame) {
  const ctx = canvas.getContext('2d');
  const { width, height } = canvas;
  const x0 = width / 2;
  const y0 = height / 2;

  ctx.save();
  ctx.resetTransform();
  places.forEach((place) => {
    const point = frame.project(place);
    const isVisible = frame.isCoordinateVisible(place);

    if (isVisible) {
      // draw point
      ctx.beginPath();
      ctx.strokeStyle = 'firebrick';
      ctx.lineWidth = 2;
      ctx.fillStyle = 'yellow';
      ctx.ellipse(point.x, point.y, 5, 5, 0, 0, 2 * Math.PI);
      ctx.fill();
      ctx.stroke();
    } else {
      // draw an arrow at the edge (y is down)
      const { x, y, direction, border } = frame.borderPosition(place);

      const arrowSize = 15;
      const x1 = x + arrowSize * Math.cos(direction + Math.PI + Math.PI / 8);
      const y1 = y + arrowSize * Math.sin(direction + Math.PI + Math.PI / 8);
      const x2 = x + arrowSize * Math.cos(direction + Math.PI - Math.PI / 8);
      const y2 = y + arrowSize * Math.sin(direction + Math.PI - Math.PI / 8);

      ctx.beginPath();
      ctx.strokeStyle = 'none';
      ctx.fillStyle = 'black';
      ctx.moveTo(x, y);
      ctx.lineTo(x1, y1);
      ctx.lineTo(x2, y2);
      ctx.closePath();
      ctx.fill();

      // arrow segment, testing offset feature
      const arrowSegmentLength = 20 + arrowSize;
      const {
        x: segmentX,
        y: segmentY,
        border: borderArrow,
      } = frame.borderPosition(place, arrowSegmentLength);
      const startX = x + (arrowSize - 3) * Math.cos(direction + Math.PI);
      const startY = y + (arrowSize - 3) * Math.sin(direction + Math.PI);

      ctx.beginPath();
      ctx.strokeStyle = 'black';
      ctx.lineCap = 'round';
      ctx.lineWidth = 3;
      ctx.moveTo(startX, startY);
      ctx.lineTo(segmentX, segmentY);
      ctx.stroke();

      // simulate inset maps with an offset of arrowSegmentLength, check that
      // they are placed on the correct border
      const colors = {
        top: 'firebrick',
        bottom: 'forestgreen',
        left: 'rebeccapurple',
        right: 'steelblue',
      };

      ctx.beginPath();
      ctx.strokeStyle = colors[borderArrow];
      ctx.lineWidth = 2;
      ctx.strokeRect(
        segmentX - arrowSegmentLength,
        segmentY - arrowSegmentLength,
        2 * arrowSegmentLength,
        2 * arrowSegmentLength,
      );

      // test offset for isCoordinateVisible around the simulated inset map
      [
        [segmentX - arrowSegmentLength, segmentY - arrowSegmentLength, 1],
        [segmentX + arrowSegmentLength, segmentY + arrowSegmentLength, -1],
      ].forEach(([startX, startY, direction]) => {
        [
          [1, 0],
          [0, 1],
        ].forEach(([dx, dy]) => {
          [0, 0.2, 0.4, 0.6, 0.8, 1].forEach((alpha) => {
            const x = startX + direction * dx * alpha * 2 * arrowSegmentLength;
            const y = startY + direction * dy * alpha * 2 * arrowSegmentLength;
            const coords = frame.unproject({ x, y });
            const visible = frame.isCoordinateVisible(coords, arrowSegmentLength);

            ctx.strokeStyle = 'none';
            ctx.fillStyle = visible ? 'green' : 'red';
            ctx.beginPath();
            ctx.arc(x, y, 2, 0, 2 * Math.PI);
            ctx.closePath();
            ctx.fill();
          });
        });
      });
    }
  });
  ctx.restore();
}
